src/bp/core/routers/admin/workspaces.ts

```diff
@@ -0,0 +1,131 @@
+import { Logger } from 'botpress/sdk'
+import { defaultPipelines } from 'common/defaults'
+import { CreateWorkspace } from 'common/typings'
+import { WorkspaceCreationSchema } from 'common/validation'
+import { ConfigProvider } from 'core/config/config-loader'
+import { InvalidOperationError } from 'core/services/auth/errors'
+import { BotService } from 'core/services/bot-service'
+import { WorkspaceService } from 'core/services/workspace-service'
+import { Router } from 'express'
+import Joi from 'joi'
+import _ from 'lodash'
+
+import { CustomRouter } from '../customRouter'
+import { NotFoundError } from '../errors'
+
+export class WorkspacesRouter extends CustomRouter {
+  constructor(
+    private logger: Logger,
+    private workspaceService: WorkspaceService,
+    private botService: BotService,
+    private configProvider: ConfigProvider
+  ) {
+    super('Workspace', logger, Router({ mergeParams: true }))
+    this.setupRoutes()
+  }
+
+  setupRoutes() {
+    const router = this.router
+    router.get(
+      '/',
+      this.asyncMiddleware(async (req, res) => {
+        res.send(await this.workspaceService.getWorkspaces())
+      })
+    )
+
+    router.post(
+      '/',
+      this.asyncMiddleware(async (req, res) => {
+        const { error } = Joi.validate(req.body, WorkspaceCreationSchema)
+        if (error) {
+          throw new InvalidOperationError(`An error occurred while creating the workspace: ${error.message}`)
+        }
+
+        await this.workspaceService.createWorkspace(req.body as CreateWorkspace)
+        res.sendStatus(200)
+      })
+    )
+
+    router.post(
+      '/:workspaceId',
+      this.asyncMiddleware(async (req, res) => {
+        const { workspaceId } = req.params
+        const { name, description } = req.body
+
+        const { error } = Joi.validate(req.body, {
+          name: Joi.string()
+            .max(50)
+            .required(),
+          description: Joi.string()
+            .max(500)
+            .allow('')
+        })
+
+        if (error) {
+          throw new InvalidOperationError(`An error occurred while updating the workspace: ${error.message}`)
+        }
+
+        await this.workspaceService.mergeWorkspaceConfig(workspaceId, { name, description })
+        res.sendStatus(200)
+      })
+    )
+
+    router.post(
+      '/:workspaceId/pipeline',
+      this.asyncMiddleware(async (req, res) => {
+        const { workspaceId } = req.params
+        const { pipelineId, resetStage } = req.body
```

Comment:
What is resetStage ? If one changes the pipeline while already built bots we should always reset the stage don't you think ? I don't mind keeping the option though

Reply:
There is still work to be done in that process. When you switch from none to botpress, bots should stay in production, but should be copied to dev with suffix "-dev" and in staging with suffix "-staging". Right now, if you reset them to "dev", then you will have all the bots without the suffix in the first stage...

I left that so you can move them around easily (from prod to dev and dev to prod), but still open to discuss for better ideas